blocks/extensions/__init__.py

```diff
@@ -141,36 +143,64 @@ class SimpleExtension(TrainingExtension):
         If not ``None``, :meth:`do` is invoked after every n-th batch.
 
     """
-    def __init__(self, before_training=False, before_first_epoch=False,
-                 on_resumption=False, on_interrupt=False,
-                 after_every_epoch=False, after_every_batch=False,
-                 after_training=False,
-                 after_n_epochs=None, after_n_batches=None,
-                 every_n_batches=None, **kwargs):
-        super(SimpleExtension, self).__init__(**kwargs)
+    BOOLEAN_TRIGGERS = frozenset(["before_training", "before_first_epoch",
+                                  "on_resumption", "on_interrupt",
+                                  "after_every_epoch", "after_every_batch",
+                                  "after_training"])
+
+    INTEGER_TRIGGERS = frozenset(["after_n_epochs", "after_n_batches",
+                                  "every_n_epochs", "every_n_batches"])
+
+    def __init__(self, **kwargs):
         self._conditions = []
-        if before_training:
-            self.add_condition("before_training")
-        if before_first_epoch:
-            self.add_condition(
-                "before_epoch",
-                predicate=lambda log: log.status.epochs_done == 0)
-        if on_resumption:
-            self.add_condition("on_resumption")
-        if on_interrupt:
-            self.add_condition("on_interrupt")
-        if after_every_epoch:
-            self.add_condition("after_epoch")
-        if after_every_batch:
-            self.add_condition("after_batch")
-        if after_training:
-            self.add_condition("after_training")
-        if after_n_epochs:
-            self.invoke_after_n_epochs(after_n_epochs)
-        if after_n_batches:
-            self.invoke_after_n_batches(after_n_batches)
-        if every_n_batches:
-            self.invoke_every_n_batches(every_n_batches)
+        super_kwargs = {}
+        trigger_keywords = self.BOOLEAN_TRIGGERS | self.INTEGER_TRIGGERS
+        for key, value in kwargs.items():
+            if key not in trigger_keywords:
+                super_kwargs[key] = value
+        self.set_conditions(**kwargs)
```

Comment:
I was actually thinking the same thing. I can imagine this being horribly
annoying source of bugs in user code.

On Fri, Feb 6, 2015 at 10:13 PM, Bart van Merriënboer <
notifications@github.com> wrote:

> In blocks/extensions/__init__.py
> <https://github.com/bartvm/blocks/pull/207#discussion_r24286296>:
>
> > -        if after_every_batch:
> > -            self.add_condition("after_batch")
> > -        if after_training:
> > -            self.add_condition("after_training")
> > -        if after_n_epochs:
> > -            self.invoke_after_n_epochs(after_n_epochs)
> > -        if after_n_batches:
> > -            self.invoke_after_n_batches(after_n_batches)
> > -        if every_n_batches:
> > -            self.invoke_every_n_batches(every_n_batches)
> > +        super_kwargs = {}
> > +        trigger_keywords = self.BOOLEAN_TRIGGERS | self.INTEGER_TRIGGERS
> > +        for key, value in kwargs.items():
> > +            if key not in trigger_keywords:
> > +                super_kwargs[key] = value
> > +        self.set_conditions(**kwargs)
>
> Just one tiny thing before we finally merge this (taking awfully long!): I
> might be overlooking something, but right now all kwargs get passed on to
> set_conditions, regardless of whether they are in trigger_keywords or
> not, right? And if it's not in trigger_keywords, it get's silently
> ignored there. Wondering whether we should raise an error somewhere in
> those cases. Otherwise, if you mistype an argument your extension won't get
> triggered and you probably won't even realize, because you didn't get an
> error.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bartvm/blocks/pull/207/files#r24286296>.
>